Reserve tids for multi-inserts.

soumyadeep2007 opened the pull request:
This solves the table bloat observed from concurrent multi-inserts.
Discussion: <Link to upstream message>

We reserve MULTI_INSERT_TID_RESERVATION_FACTOR times the number of tids
requested in a multi-insert call. This ensures that we don't have the
inefficient page splits described in the thread.

For large values of MULTI_INSERT_TID_RESERVATION_FACTOR, the number of
tid items created sometimes exceeds 20 (XLR_NORMAL_RDATAS). Thus we need
to call XLogEnsureRecordSpace to reserve enough rdatas before we call
zsbt_wal_log_tidleaf_items.

Also, I discovered that our estimation of internal pages resultant from
a split may be off by one. So, I relaxed the assert for now.

Discussion: https://www.postgresql.org/message-id/CADwEdopF2S6uRXJRg%3DVZRfPZis80OnawAOCTSh_SrN2i1KGkMw%40mail.gmail.com

soumyadeep2007 committed:
Reserve tids for multi-inserts.

This solves the table bloat observed from concurrent multi-inserts.
Discussion:
https://www.postgresql.org/message-id/CAAKRu_YxyYOCCO2e83UmHb51sky1hXgeRzQw-PoqT1iHj2ZKVg%40mail.gmail.com

We reserve MULTI_INSERT_TID_RESERVATION_FACTOR times the number of tids
requested in a multi-insert call. This ensures that we don't have the
inefficient page splits described in the thread.

soumyadeep2007 commented:
Analysis of solution with numbers here:
https://greenplum.slack.com/files/UH4TLR1T3/FTR089C00/zedstore_performance_txt.txt
I will be turning this into an upstream message.

ashwinstar commented on src/backend/access/zedstore/zedstore_tidpage.c:
```diff
@@ -1380,6 +1380,17 @@ zsbt_tid_add_items(Relation rel, Buffer buf, List *newitems, zs_pending_undo_op
+			 * This must be called before we enter the critical section as
+			 * XLogEnsureRecordSpace performs memory allocation.
+			 */
+			XLogEnsureRecordSpace(0, list_length(newitems) + 1 + 2);
```
I would like if you can spit these changes into separate commit. They are not directly connected with reserve tids, but are required in-general with prior code as well.

soumyadeep2007 replied:
Makes sense!

ashwinstar commented on src/backend/access/zedstore/zedstore_tidpage.c:
```diff
@@ -1915,6 +1940,10 @@ zsbt_binsrch_tidpage(zstid key, Page page)
 }
 
+/*
+ * It must be called after XLogBeginInsert and XLogEnsureRecordSpace have been
```
Any assertion we can add to validate `XLogEnsureRecordSpace()` was called when we get to this function?

soumyadeep2007 replied:
Unfortunately, the globals that `XLogEnsureRecordSpace()` and `XLogInsertInit()` set are static and not exposed. Thus, we can't really assert that they have been called.

ashwinstar commented on src/backend/access/zedstore/zedstore_tuplebuffer.c:
```diff
@@ -19,17 +19,18 @@
 
-#define TID_RESERVATION_SIZE		100
+#define SINGLE_INSERT_TID_RESERVATION_SIZE		100
+#define MULTI_INSERT_TID_RESERVATION_FACTOR		10
```
Having comment explaining these two would be good.

soumyadeep2007 replied:
Will add some!

ashwinstar commented on src/backend/access/zedstore/zedstore_tuplebuffer.c:
```diff
@@ -138,10 +137,27 @@ get_tuplebuffer(Relation rel)
+ * For multi-inserts, we request and reserve (ntids * MULTI_INSERT_TID_RESERVATION_FACTOR)
+ * number of tids.
+ *
+ * TODO: expose these as GUCs as they are very workload sensitive (especially
```
Should this TODO be moved to where `MULTI_INSERT_TID_RESERVATION_FACTOR` is defined

soumyadeep2007 replied:
Makes sense!

ashwinstar commented on src/backend/access/zedstore/zedstoream_handler.c:
```diff
@@ -184,7 +184,7 @@ zedstoream_insert_internal(Relation relation, TupleTableSlot *slot, CommandId ci
 
 	if (speculative_token == INVALID_SPECULATIVE_TOKEN)
-		tid = zsbt_tuplebuffer_allocate_tid(relation, xid, cid);
+		tid = zsbt_tuplebuffer_allocate_tids(relation, xid, cid, 1);
```
What does `1` mean here?

soumyadeep2007 replied:
1 is the number of tids we want to allocate.

ashwinstar reviewed (approved):
Looks was fun working on this! Feel free to merge once you think is done.

soumyadeep2007 commented:
Thank you so much for the review Ashwin! I will incorporate your changes and merge it in!

soumyadeep2007 committed:
Relax assert for estimated number of internal pages

There are situations where the number of internal pages resultant from a
split may be off by one as compared to our estimated cxt.num_pages. So
relaxing the assert for now.

soumyadeep2007 committed:
Reserve rdatas for WALing tid tree leaf items

Inserting a large number of tids into the tid tree at once may cause
zsbt_wal_log_tidleaf_items() to be called with more than
XLR_NORMAL_RDATAS number of tid items. This leads to:
ERROR: "too much WAL data"
when XLogRegisterBufData() is called from within
zsbt_wal_log_tidleaf_items().

To remedy this situation ensure that there is enough rdatas allocated
for all of the items and for any associated undo_op.
